compiler/sem/semstmts.nim

```diff
@@ -1080,14 +1080,26 @@ proc semNormalizedConst(c: PContext, n: PNode): PNode =
       c.p.owner.kind != skMacro:
     typFlags.incl taProcContextIsNotMacro
 
-  def = semConstExpr(c, def)
-
-  if def.kind != nkNilLit:
-    # xxx: this seems wrong, instead of guarding on nil, maybe it should be
-    #      inferred from `typ`? then we still do the rest of the checks?
-    if c.matchedConcept != nil:
-      typFlags.incl taConcept
-    typ = typeAllowedOrError(typ, skConst, c, def, typFlags)
+  if c.matchedConcept != nil:
+    typFlags.incl taConcept
+
+  # semantically analyse the expression first, so that it can be rejected
+  # early based on the type before trying to evaluate it
+  def = semExprWithType(c, def)
+  typ = typeAllowedOrError(typ, skConst, c, def, typFlags)
+
+  # XXX: ``fitNode`` is never used here. That doesn't seem correct, as the
+  #      type of the expression is used, no matter what
+  if not typ.isError:
+    let e = evalConstExpr(c, def)
+    if e.isError:
+      # XXX: an error node can't be used as the def for a ``const``, as error
+      #      nodes are not considered when inlining constants. If there occurrs
+      #      an error during analysis of the ``const``, the resulting symbol
+      #      should likely be an ``skError`` symbol
+      localReport(c.config, e)
```

Comment:
If we keep the error and not report it eagerly then we can propagate that into the productions below and register an skError.

Reply:
Yep, but no `skError` symbol is produced right now. Since the logic in `semNormalizedConst` is fairly complex and this PR already does a good amount of other things besides refactoring `semConstExpr`, my thinking was that it'd be better to eagerly report the error for now.